File: LeetCode/17. Letter Combinations of a Phone Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits == "") return {};

        vector<string> ans {""};
        vector<string> mapping {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        
        for (auto digit : digits) {
            vector<string> temp;

            for (auto prevChars : ans) {
                for (auto currChars : mapping[digit - '2']) {
                    temp.push_back(prevChars + currChars);
                }
            }

            ans = temp;
        }
        
        return ans;
    }
};

File: LeetCode/17. Letter Combinations of a Phone Number.rb
# @param {String} digits
# @return {String[]}
def letter_combinations(digits)
    return [] if digits.size == 0
    
    ans = [""]
    mapping = ["abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
    
    digits.each_char do |digit|
        temp = []
        ans.each do |prevChar|
            mapping[digit.ord - '2'.ord].each_char do |currChar|
                temp << prevChar + currChar
            end
        end
        ans = temp
    end
    
    return ans
end
